osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -55,7 +55,12 @@ private double calculateRhythmBonus(DifficultyHitObject current)
 
             bool firstDeltaSwitch = false;
 
-            for (int i = Previous.Count - 2; i > 0; i--)
+            int rhythmStart = Math.Min(Previous.Count - 2, 0);
```

Comment:
Dunno if it's just me but I find the usage of min rather here rather confusing, especially so that the variable is next used to index `Previous` in the while loop, which looks uber dangerous given `rhythmStart` is sometimes negative - it just so happens negative values are never used because of two disparate failing branches (`rhythmStart < Previous.Count - 2` in the while loop and `i > 0` in the for loop, respectively).

I'd be much more comfortable with an early return guard to get that edge case out of the way as early as possible:

```csharp
if (Previous.Count < 2)
    return 0;
```

Reply:
Your explanation of the index safety guarantee is exactly correct. For me it's obvious that the index is safe even if Previous.Count-2 results in a negative index; but I can see how this is likely from the bias of me being the author.

Early returning 0 if Previous.Count is below 2 would change the results of the code because of the constant 4 in the tail ended return. Returning 2 or even sqrt(4) would be a bit obtuse would it not?

Would it be applicable to surround the entire loop in a new scope with a (Previous.Count < 2) predicate? Personally I feel the extra scoping leads to it being a little more obtuse to read.

Reply:
Fair point about the constant. In that case I'd probably ship off the entire calculation of `rhythmComplexitySum` to a separate method, which makes it possible to early-return 0 as proposed above. However, since this is obviously somewhat subjective and there's a difference in opinion, I'll wait for another opinion on this matter.